### ISSUE TITLE
5x speedups without JIT compiler + portability...

### ISSUE THREAD

**Issue by sbrunthaler** [faster-cpython/ideas#93]
Hi guys, 

one of my students just pointed me at your repository and I just wanted to briefly touch base and highlight some of my own prior work/research in optimizing the CPython interpreter. Pretty much ten years ago, I did extensive research on purely-interpretative optimizations, mostly through inline caching with quickening. Subsequent research on combining multiple different techniques led to maximum speedups of 5.5x, *without* requiring a JIT compiler. AFAICT, my research would ideally suited for stages 1 and 2 of your implementation plan. Due to some interest expressed on this research on Twitter about three months ago, I put together the paper and the many rejections it got in academia. If you're interested, please take a look: https://arxiv.org/abs/2109.02958

Based on my experience, it should be possible to obtain much of the proposed speedups by focusing just on the interpreter-based optimizations. A simple JIT is, IMHO, not going to provide more speedups beyond the interpreter. (Similar to having a template JIT vs an optimizing interpreter, as the template JIT mostly targets instruction dispatch costs, which are usually not dominant costs in Python.) A nice benefit is that this strategy maintains portability, and could even be used to provide optimizations across extensions written in C (numpy, etc.).

If I can provide any further explanations beyond the paper, please do let me know!

Other than that: Have a nice day and weekend, respectively & all the best from Munich,
--stefan

**Comment by sbrunthaler** [comment]
I've seen the "old" patch set from 2012. AFAIR, this is mostly just inline caching of interpreter instructions, including some nifty caching of LOAD_GLOBAL/LOAD_ATTR. The link to the new paper (was unfortunately never accepted anywhere) does some more aggressive optimizations, which is where the speedups come from. 
* First, it also maps Python objects to native-machine data types and rewrites longer sequences of CPython interpreter instructions to heavily optimized interpreter instructions, similar to how a JIT compiler would optimize CPython, but thanks to the purely-interpretative nature, still portable. One cool side effect of optimizing CPython like this is that retains full compatibility with existing reference-counting semantics (also modules/extensions written in C), while being able to *reduce* memory consumption for hot paths. IIRC in spectralnorm, my interpreter used *less* memory than standard CPython.
* Second, another 1x speedup improvement comes from using native-machine type based superinstructions: The longer sequences of type and data representation optimized interpreter instructions have so little overhead in their corresponding interpreter instruction implementations that they finally end up suffering from instruction dispatch costs. I wrote a small recorder for the optimized sequences, which can be used to generate new superinstructions pretty easily. Those superinstructions can be baked into the CPython interpreter at interpreter compile-time and are highly optimized by the ahead-of-time compiler (being able to produce heavily optimized native-machine code).

In combination, both techniques allow for substantial speedups. Of course this depends also on the speedup potential of the relevant benchmarks. In the linked paper, I also did some analysis to approximate the overhead of interpretation on specific programs and found that the technique eliminates most of this overhead. Even doing this analysis on prospective benchmarks a priori would be highly relevant, as this serves as a good guideline to how good the optimizations are performing.

If I can be of any help to your efforts, let me know. Also happy to do a VC!
